### PR TITLE
Feature/async

### PR DESCRIPTION
**Pull request by eeerin**
now using ES6.

this seems like a good stopping point, before removing the Immutable stuff.

### CODE REVIEW

File: lib/db.js [hunk 1]
```diff
@@ -1,338 +1,270 @@
...
-    this._db = db;
     this._connectionStatus = 'connected';
-    this._db.on('close', function () {
+    this._db.on('close', () => {
       debug('database connection closed');
       this._db = null;
       this._connectionStatus = 'disconnected';
       this.emit('disconnect');
-    }.bind(this));
-    this.emit('connect', this);
```
Comment (spacetc62) [inline lib/db.js]: looks like these two lines got accidentally removed

File: lib/db.js [hunk 2]
```diff
@@ -1,338 +1,270 @@
...
-      deferred.resolve(collection);
     });
-  } else if (this._connectionStatus === 'connecting') {
-    debug('waiting on connection before loading collection ' + name);
-    this.on('connect', function () {
-      this._db.collection(name, options, function (err, collection) {
-        if (err) { return deferred.reject(err); }
-        debug('collection ' + name + ' loaded');
-        deferred.resolve(collection);
+    this._db.collection = promisify(this._db.collection);
```
Comment (spacetc62) [inline lib/db.js]: you don't need to promisfy this guy, it just directly returns collection

File: lib/db.js [hunk 3]
```diff
@@ -1,338 +1,270 @@
...
+   */
+  async collection(name, options) {
+    if (this._connectionStatus === 'connected') {
+      debug(`loading collection ${name}`);
+      const collection = await this._db.collection(name, options);
+      debug(`collection ${name} loaded`);
+      return collection;
+    } else if (this._connectionStatus === 'connecting') {
+      debug(`waiting on connection before loading collection ${name}`);
+      this.on('connect', async () => {
```
Comment (spacetc62) [inline lib/db.js]: need to do a promise thingy here and actually return collection

File: lib/model.js [hunk 4]
```diff
@@ -315,247 +215,243 @@ function _create (data, definition) {
...
+      return db.distinct(collection, key, query, options);
+    },
+
+    aggregate: (pipeline, options) => {
+      const collection = definition.collection;
+      debug(`aggregating ${JSON.stringify(pipeline)} from ${collection}`);
+      return db.aggregate(collection, pipeline, options);
+    },
+
+    project: (query, value, options) => {
```
Comment (spacetc62) [inline lib/model.js]: not async-ing this guy?

File: lib/model.js [hunk 5]
```diff
@@ -315,247 +215,243 @@ function _create (data, definition) {
...
-    }.bind(this));
+  definition.methods.remove = async function() {
+    const collection = definition.collection;
+    const query = { _id: this._id, _etag: this._etag };
+    const before = async () => {
+      if ('function' === typeof this.beforeRemove) {
+        return this.beforeRemove();
+      }
+    };
+    await before();
```
Comment (spacetc62) [inline lib/model.js]: why not just:
```
if ('function' === typeof this.beforeRemove) {
  await return this.beforeRemove();
}
```

File: lib/model.js [hunk 6]
```diff
@@ -315,247 +215,243 @@ function _create (data, definition) {
...
-    if ('function' === typeof this.beforeSave) {
-      before = q.fcall(() => { return this.beforeSave.apply(obj); })
-        .then(function () { return obj; });
+    let obj = this.toObject();
+    const before = async () => {
+      if ('function' === typeof this.beforeSave) {
+        return this.beforeSave.apply(obj);
+      }
+    };
+    await before();
```
Comment (spacetc62) [inline lib/model.js]: same thing as my comment on the before remove

File: lib/model.js [hunk 7]
```diff
@@ -315,247 +215,243 @@ function _create (data, definition) {
...
+    debug(`removing ${JSON.stringify(query)} in ${collection}`);
+    const result = await db.remove(collection, query);
+    if (result.deletedCount === 0) {
+      throw errors.Conflict({ message: 'Document remove conflict' });
+    }
+    debug(`remove from ${collection} successful`);
+    if ('function' === typeof this.afterRemove) {
+      // should this be awaited??
```
Comment (spacetc62) [inline lib/model.js]: think we decided on no, and should probably have an explicit comment to that effect?

File: lib/model.js [hunk 8]
```diff
@@ -249,46 +163,32 @@ function _create (data, definition) {
...
+    const original = this;
+    return this.mutate(function() {
       objectPath.del(this, path);
       _markAsModified(this, [path], original);
     });
   };
 
-  definition.methods.has = function (path) {
+  definition.methods.has = function(path) {
+    // TODO use omnibelt path
```
Comment (spacetc62) [inline lib/model.js]: probably not, because subtle differences

File: lib/model.js (hunk 5 above)
Comment (eeerin) [reply lib/model.js]: cause brainfart...

**Comment by spacetc62** [comment]
@eeerin how is it going with making this PR work with the models tests?

**Comment by eeerin** [comment]
most of the tests are just removing q as a dependency, I'm about halfway complete. I should probably stop trying to change them all into async tests.

**Pull request merged**